### PR TITLE
Zypper auth

### PR DESCRIPTION
**Pull request by rjschwei**

### CODE REVIEW

File: usr/lib/zypp/plugins/urlresolver/susecloud [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import cloudregister.registerutils as utils
+
+from zypp_plugin import Plugin
+
+utils.start_logging()
+utils.set_proxy()
+
+
+class SUSECloudPlugin(Plugin):
+    def RESOLVEURL(self, headers, body):
```
Comment (smarlowucf) [inline usr/lib/zypp/plugins/urlresolver/susecloud]: Consider snakecase `resolve_url`.

File: usr/lib/zypp/plugins/urlresolver/susecloud [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        repo_url = srv_url + headers.get('path')
+        credentials = headers.get('credentials')
+        if credentials:
+            repo_url += '?credentials=' + credentials
+        instance_data = bytes(
+            utils.get_instance_data(utils.get_config()),
+            'utf-8'
+        )
+        self.answer(
+            'RESOLVEDURL',
```
Comment (smarlowucf) [inline usr/lib/zypp/plugins/urlresolver/susecloud]: Typo?

File: usr/sbin/cloudguest-repo-service [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    for repo in product_info.get('repositories'):
+        name = repo.get('name')
+        if name not in available_repos:
+            continue
+        refresh = repo.get('autorefresh')
+        enabled = repo.get('enabled')
+        url = repo.get('url')
+        if url.startswith('plugin:'):
+            is_plugin_url = True
+            with open(service_revert_file, 'a') as trigger:
```
Comment (smarlowucf) [inline usr/sbin/cloudguest-repo-service]: :+1: for context managers.

File: usr/sbin/cloudguest-repo-service [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+# Get the available repos from the server
+user, password = utils.get_credentials(
+    utils.get_credentials_file(update_server)
+)
+auth_creds = HTTPBasicAuth(user, password)
+instance_data = bytes(utils.get_instance_data(utils.get_config()), 'utf-8')
+headers = {}
+if instance_data:
+    headers['X-Instance-Data'] = base64.b64encode(instance_data)
+req = requests.get(
```
Comment (smarlowucf) [inline usr/sbin/cloudguest-repo-service]: Ditto req is a response which is confusing name.

File: lib/cloudregister/registerutils.py [hunk 5]
```diff
@@ -391,7 +521,56 @@ def get_repo_url(repo_name):
...
 
-    return None
+    return ''
+
+
+# ----------------------------------------------------------------------------
+def get_smt():
+    """Returns an update server that is reachable."""
+
+    smt_ip = None
```
Comment (smarlowucf) [inline lib/cloudregister/registerutils.py]: Not used?

File: lib/cloudregister/registerutils.py [hunk 6]
```diff
@@ -391,7 +521,56 @@ def get_repo_url(repo_name):
...
+                        '"%s"' % str(
+                            (new_target.get_ipv4(), new_target.get_ipv6())
+                        )
+                    )
+                    replace_hosts_entry(current_smt, new_target)
+                    set_as_current_smt(new_target)
+                    return new_target
+    else:
+        # Try any other update server we might know about
+        for smt in available_servers:
```
Comment (smarlowucf) [inline lib/cloudregister/registerutils.py]: Shadows import from https://github.com/SUSE-Enceladus/cloud-regionsrv-client/pull/3/files#diff-ee2b1d378707c9bf0b0c65a507795a1eR31

File: lib/cloudregister/registerutils.py [hunk 7]
```diff
@@ -569,6 +767,53 @@ def set_proxy():
...
+    for update_server in update_servers:
+        service_targets.append(update_server.get_ipv4())
+        service_targets.append(update_server.get_ipv6())
+        service_targets.append(update_server.get_FQDN())
+
+    for service_file in service_files:
+        link_created = None
+        cfg = configparser.RawConfigParser()
+        try:
+            parsed = cfg.read(service_file)
```
Comment (smarlowucf) [inline lib/cloudregister/registerutils.py]: Parsed never used.

File: usr/lib/zypp/plugins/urlresolver/susecloud (hunk 1 above)
Comment (ikapelyukhin) [reply usr/lib/zypp/plugins/urlresolver/susecloud]: Zypper resolver plugin entry point has to be named `RESOLVEURL`.

File: usr/lib/zypp/plugins/urlresolver/susecloud (hunk 2 above)
Comment (ikapelyukhin) [reply usr/lib/zypp/plugins/urlresolver/susecloud]: No, that's what zypper resolver plugin has to respond with :-)

File: lib/cloudregister/registerutils.py [hunk 8]
```diff
@@ -271,6 +278,46 @@ def find_repos(contains_name):
     return repo_names
 
 
+# ----------------------------------------------------------------------------
+def get_activations():
+    """Get the activated products from the update server"""
+    activations = {}
+    update_server = get_smt()
+    user, password = get_credentials(get_credentials_file(update_server))
+    if not user and password:
```
[the remark this answers is not in the document]
Comment (ikapelyukhin) [reply lib/cloudregister/registerutils.py]: Probably `not (user and password)`.

File: lib/cloudregister/registerutils.py [hunk 9]
```diff
@@ -337,11 +441,32 @@ def get_current_smt():
     return smt
 
 
+# ----------------------------------------------------------------------------
+def get_installed_product_names():
```
Comment (ikapelyukhin) [inline lib/cloudregister/registerutils.py]: It seems that this function isn't used anywhere.

File: lib/cloudregister/registerutils.py [hunk 10]
```diff
@@ -378,8 +503,13 @@ def get_instance_data(config):
                     logging.error(errMsg)
 
     if instance_data:
+        # Marker for the server to not return https:// formated
+        # service and repo information
+        instance_data += b'<repoformat>plugin:susecloud</repoformat>\n'
         return instance_data.decode()
 
+    return instance_data
```
Comment (ikapelyukhin) [inline lib/cloudregister/registerutils.py]: Seems like `.decode()` got lost here. Not sure what encoding it has. Also now there are two `return instance_data` statements -- one should suffice.

File: lib/cloudregister/registerutils.py [hunk 11]
```diff
@@ -595,12 +840,19 @@ def replace_hosts_entry(current_smt, new_smt):
     """Replace the information of the SMT server in /etc/hosts"""
     known_hosts = open(HOSTSFILE_PATH, 'r').readlines()
     new_hosts = ''
+    if not current_smt:
+        logging.error(
+            'System in inconsistent state, request to replace '
+            'entry in hosts file, but not registration server '
+            'provided. WIll not take any action'
```
Comment (ikapelyukhin) [inline lib/cloudregister/registerutils.py]: * `not registration server` -> `no registration server`
* `WIll` -> `Will`

File: usr/lib/zypp/plugins/urlresolver/susecloud [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        if not update_server:
+            # Something went seriously wrong, however try the name from
+            # the hosts file
+            server_name = utils.get_update_server_name_from_hosts()
+        else:
+            server_name = update_server.get_FQDN()
+        if not server_name:
+            # Point to localhost, which is wrong but produces a valid url
+            # zypper will fail gracefully with a reasonable message
+            server_name = '127.0.0.1'
```
Comment (ikapelyukhin) [inline usr/lib/zypp/plugins/urlresolver/susecloud]: That seems unnecessary. If the instance is running a web server, zypper will make requests to it (with instance data in the headers).

`self.answer('RESOLVEDURL', {}, None)` seems to result in the following error:

```
Repository 'Test' is invalid.
[Test|plugin:/susecloud?test=1] Valid metadata not found at specified URL
```
That seems good enough?

File: lib/cloudregister/registerutils.py [hunk 13]
```diff
@@ -271,6 +278,46 @@ def find_repos(contains_name):
...
+        return activations
+
+    auth_creds = HTTPBasicAuth(user, password)
+
+    instance_data = bytes(get_instance_data(get_config()), 'utf-8')
+    headers = {}
+    if instance_data:
+        headers['X-Instance-Data'] = base64.b64encode(instance_data)
+
+    req = requests.get(
```
[the remark this answers is not in the document]
Comment (rjschwei) [reply lib/cloudregister/registerutils.py]: Cleaned up

**Comment by rjschwei** [comment]
@ikapelyukhin @smarlowucf I think all comments should be addressed in one way or another.

**Pull request merged**